Support odk:setlocation action

lognaturel opened the pull request:
Closes #3174 

This implementation always uses the Google fused location API configured to balance power and accuracy. For each action triggered, locations are acquired for 20s and the highest-accuracy reading is saved. The only exception to that is if the action is in a repeat and gets triggered again, the timer resets for each repeat instance. 

ANY failure to get location leaves the field blank (unlike with the audit feature, there is no way to know what went wrong).

#### What has been done to verify that this works as intended?
launch form with odk:setlocation action: 
- [x] when Collect does not have device-level location permissions, verify that permissions are requested (note - if location acquisition is triggered by form open, it won't be acquired that first time, see below for explanation)
- [x] verify that location is not saved if device-level permission is not given
- [x] verify that snackbar letting user know of location collection is shown on every form open (from new blank instance, editing existing instance, recovery from savepoint)
- [x] triggered by form open, verify that location is written to the instance
- [x] triggered by value change, verify that location is written on value change
- [x] triggered by value change in repeat, verify that location is written on value change for each repeat instance
- [x] verify that location preference toggle is available from overflow menu
- [x] toggle location preference off, verify that location is not collected
- [x] toggle location back on, verify that location is collected
- [ ] verify that location client is stopped even if form is closed before 20 seconds have gone by

#### Why is this the best possible solution? Were any other approaches considered?
I initially considered using the existing location client in `FormEntryActivity` that the audit location implementation uses. That proved problematic for two major reasons:
- the action handlers are registered into a static map in `XFormParser` and exist as long as the application is running with no way to unregister them (see `FormController.initializeJavaRosa`). As far as I know, this limits our options for per-activity or per-form behavior. The only way I could think to do that is with a general broadcast using something like an event bus.
- managing when each setlocation action and the audit log are requesting locations is tricky. It would require maintaining state about which object started location requests when.

This means that there is no permissions checking when the location client is actually used, it's all handled by the activity after the form has been loaded and before the action is triggered. An unfortunate implication of this is that if location permissions aren't already granted the first time a form with a setlocation action triggered by form open is launched, the location won't be captured. This is because the permissions check will come after the form is loaded. I think it's minor enough to be ok.

Another decision point was how much to try to pull logic out of the `FormEntryActivity`. I ended up pulling everything related to background location out.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
- the audit location logging feature got a lot of modifications and is at risk, especially around lifecycle-management edge cases
- I pulled out code to format locations to the geopoint format so that affects geopoint
- there's a risk of resource leaks because setlocation actions exist for the whole time a form is set as the current form in the form controller. When a new form is loaded, the old global form controller is just set to a new reference.

#### Do we need any specific form for testing your changes? If so, please attach one.
[setlocation-action-instance-load.xml.txt](https://github.com/opendatakit/collect/files/3308672/setlocation-action-instance-load.xml.txt)
[setlocation-action-value-changed.xml.txt](https://github.com/opendatakit/collect/files/3384139/setlocation-action-value-changed.xml.txt)
[setlocation-action-with-repeats.xml.txt](https://github.com/opendatakit/collect/files/3308674/setlocation-action-with-repeats.xml.txt)
[setlocation-and-audit-location.xml.txt](https://github.com/opendatakit/collect/files/3374174/setlocation-and-audit-location.xml.txt)

#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/opendatakit/docs/issues/new) and include the link below.

Yes, but we'll need pyxform support, first.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

lognaturel committed:
Add support for odk:setlocation action

lognaturel committed:
Ignore Visual Studio Code files

lognaturel committed:
Show the location options menu when formdef has setlocation

lognaturel committed:
Perform location checks when form has setlocation action

lognaturel committed:
Only log location-related events if location auditing is enabled

lognaturel committed:
Pull location updating logic out and test it

lognaturel committed:
Pull out location provider change receiver

lognaturel committed:
Clarify onStart case

lognaturel committed:
Add basic UI tests

lognaturel committed:
Pull location-related code out of FormEntryActivity, test it

lognaturel commented:
I didn't have enough confidence that all the condition checking and user feedback were right in all cases without tests so I added 388b704. Thanks also to @seadowg for the gentle nudge to Do The Right Thing™️. @grzesiek2010, @seadowg and anyone else interested, I'd love some preliminary feedback on it. Is it too aggressive to do here? Are the tests meaningful and complete? Is the general state-machine-like approach appealing? There are still a few weird bits like `FormEntryActivity.displayUIFor`.

I'm also interested in whether you think more UI tests such as in 8a68e0e would be of value. I could probably add more with mocks but with the tests introduced in 388b704, I don't feel as much need to.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/FormEntryViewModel.java:
```diff
@@ -0,0 +1,49 @@
+    @NonNull
+    private final BackgroundLocationManager locationManager;
+
+    public FormEntryViewModel() {
```
How'd you feel about the `BackgroundLocationManager` being a constructor level dependency here? I think it would make it easier to pull apart/test in future.

lognaturel replied:
Thanks for bringing that up, I should have called it out explicitly. One of my design goals was to make the view model layer so thin that it wouldn't need its own testing. I see its role as maintaining model objects through the activity lifecycle and passing main thread actions to model objects. In a case where there might be live-updating data (form lists!), it could act as a listener for those model objects and present a livedata object to the activity. 

That way, we don't have to worry about constructor dependencies, factories, creating the dependencies at the right time so they don't get recreated on orientation changes, etc. Does that make sense? Or is there a way to manage constructor-level dependencies that is simpler?

seadowg replied:
So if we use a `ViewModel` factory to construct the view model we will actually get the same view model instance between orientation changes so we should be ok.

I think this is something I might just be overly nitpicky about but I generally like having anything your object depends on come in at a constructor level to make it easier to comprehend what things that object is going to interact with. It's also a refactor you're pretty much always going to have to carry out if you get the object to a point that you feel you'd like unit tests for it.

lognaturel replied:
I agree that constructor-level dependencies are best. Is 2e61ad2 what you had in mind? Now the activity builds its own factory dependency. I think that's ok because the activity is responsible for building many of its own dependencies already anyway. If this is not what you had in mind, do you want to give me a commit to look at?

If you saw a previous confused comment, for a minute it looked like `create` was being called multiple times and I couldn't figure out the point of having a factory. 🥴

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/FormEntryViewModel.java:
```diff
@@ -0,0 +1,49 @@
+                new BackgroundLocationHelper());
+    }
+
+    public void formFinishedLoading() {
```
I feel like a nice next step (not for now) here would be to have the ViewModel (or something it owns) coordinate form loading from disk. Probably a whole kettle of fish we want to throw out the kitchen right now (just go with me on that metaphor).

lognaturel replied:
yes, Yes, YES! And I think a state-machine-like approach would work very nicely for that as well, again keeping the ViewModel extremely thin and having another model object that keeps track of what work has been done to build up the form controller, what errors have come up, what needs to be displayed to the user, etc.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java:
```diff
@@ -0,0 +1,284 @@
+    @NonNull
+    private BackgroundLocationHelper helper;
+
+    public BackgroundLocationManager(LocationClient locationClient, BackgroundLocationHelper helper) {
```
The way `BackgroundLocationHelper` is being used here really reminds me of the ["Delegate" pattern](https://en.wikipedia.org/wiki/Delegation_pattern). Would it be too nerdy to call it a `BackgroundLocationDelegate` instead?

On that, it could be cool to have the `BackgroundLocationManager` depend on an interface rather than a concrete class. That might help make it clearer where the separation is.

lognaturel replied:
It's kind of a wacky delegate because it delegates to different objects/classes, right? It's more of a forward-er? Bag-o-methods?

> That might help make it clearer where the separation is.

I don't know what the separation is. 😭 I see `BackgroundLocationHelper` as a messy "in-progress" kind of thing. I don't really think its interface makes any sense as-is. Again, it's just a bag of stuff I needed. Or do you see a reasonable interface hiding in there? I suppose it could be something like `BackgroundLocationPreconditionChecker` and the last three methods could find another home?

I think ideally there would be a way to get to the form controller from the `BackgroundLocationManager`. It currently doesn't get created until the async task launched by the `FormEntryActivity`'s `onCreate` completes. One option might be to pass it in to the `formFinishedLoading` method? I'm actually kind of liking that... then the top three methods could go somewhere else. Another option would be to make `FormController` a true singleton and mutate it rather than throwing it away and recreating it every time a new form is loaded. Don't love that.

seadowg replied:
Yeah I think it would be nice to break into a couple of different interfaces but it might just be good first step for it to be one interface/delegate. Then IMO it's easier to look at it without the implementation and think "how should this actually work"?

lognaturel replied:
Sorry, I'm not really clear on this. Do you mean just taking all the methods in `BackgroundLocationHelper` and putting them in an interface? If not, do you want to give me a commit?

seadowg replied:
I think actually you're right looking at it again. I think a nice next step (outside the PR) would be to split it into a few different interfaces (as you're kinda saying).

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java:
```diff
@@ -0,0 +1,284 @@
+    }
+
+    public void formFinishedLoading() {
+        if (currentState == BackgroundLocationState.NOT_APPLICABLE) {
```
Maybe diving to deep into the state machine pattern but to me it would feel more consistent to have all these methods use a `switch`. In my head that makes it easy to skim through and see what states the method cares about and what it does in the default case.

lognaturel replied:
Agreed, will update. At a high-level, are you upset by the fact that the state transitions happen based not only on actions but also on various state checks triggered by the actions? For example, a more pure state machine design here might be to have the type of form checked outside this class and the event be something like `backgroundLocationFormFinishedLoading`. 

There might then also be a series of states between `APPLICABLE` (to be renamed) and `COLLECTING_LOCATION` instead of all the checks in `activityDisplayed` and `locationPermissionsGranted`.

I feel like what I ended up with strikes a good balance between grouping expected checks together and reaping the benefits of a state-machine-like structure but would like feedback on that.

seadowg replied:
Hmmm. I think it's probably good enough for now as what it does nicely is pull knowledge of those transitions away from the Activity which doesn't care about them.

lognaturel replied:
👍 And I think there's a reasonable way to describe this model if we want to follow it elsewhere -- introduce events/state transitions for changes that only the activity can detect and push everything else into the model objects.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java:
```diff
@@ -0,0 +1,284 @@
+        NOT_APPLICABLE,
+
+        /** The current form tracks background location and a message hasn't been shown to the user **/
+        APPLICABLE,
```
The naming of `APPLICABLE` and `NOT_APPLICABLE` took me a little while to get my head around. Maybe that's just me but could it be `FORM_NEEDS_LOCATION` or something?

lognaturel replied:
How about `NO_BACKGROUND_LOCATION_NEEDED` and `PENDING_PRECONDITION_CHECKS`?

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/location/LocationTestUtils.java:
```diff
@@ -0,0 +1,26 @@
+        location.setLatitude(lat);
+        location.setLongitude(lon);
+        location.setAltitude(alt);
+        return location;
```
can't this method just return `createLocation(provider, lat, lon, alt, 0);`?

lognaturel replied:
That wouldn't be the same because then the accuracy would be set. But you're right that I could call this method from the other variant and add the accuracy. Thanks, I'll do that.

lognaturel replied:
b353dac

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/logic/AuditEvent.java:
```diff
@@ -40,24 +40,43 @@
+        private final boolean isLocationRelated;
 
-        AuditEventType(String value) {
+        AuditEventType(String value, boolean isLocationRelated) {
```
I think we should fallow the same approach we use to determine if it's an internal event. We don't add a new parameter to the constructor but we have a special method: https://github.com/opendatakit/collect/pull/3175/files#diff-b25f62214ddaa8161b77f62ab938dcaeR122

It would be better to have similar `isLocationAuditEventType` method

lognaturel replied:
I agree it would be good to group everything to do with audit event types together. Ideally those would be properties on the event type entity. My preference would be to add an `isInterval` boolean property to the `AuditEventType` enum. Why would you prefer to have methods at the `AuditEvent` level only?

grzesiek2010 replied:
Ok sounds fine, I just want both cases to be consistent so we can fallow your approach.

lognaturel replied:
88c7949

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/utilities/AuditEventLogger.java:
```diff
@@ -62,41 +62,40 @@ public void logEvent(AuditEvent.AuditEventType eventType, boolean writeImmediate
-
-        // Ignore beginning of form and repeat events
-        if (eventType == AuditEvent.AuditEventType.BEGINNING_OF_FORM || eventType == AuditEvent.AuditEventType.REPEAT) {
+        if (!isAuditEnabled()
```
I would factor out this list of conditions to a separate method.

lognaturel replied:
What would that method be called? What do you think the advantage would be? All I can think of in this case would be something like `isEventLoggingRequired` but that makes understanding the `logEvent` method harder because of the added level of indirection that provides no information on when logging is unnecessary.

grzesiek2010 replied:
I was thinking about `shouldEventBeLogged` but a method with the same name already exists in this class...
advantage? If you can factor out a part of a method giving a meaningful name (to the new one) it makes the code more readable you know it maybe even better.

>that makes understanding the logEvent method harder because of the added level of indirection that provides no information on when logging is unnecessary.

if a method has a good name and (or) a comment I don't think it's less clear.

lognaturel replied:
What do you think of 2e61ad2?

grzesiek2010 replied:
What does it have to do with our conversation? I think you wanted to add it somewhere else right?

lognaturel replied:
Hmm, looks like I copied the wrong hash. 82d5c98 is what I meant.

grzesiek2010 replied:
Looks fine.

grzesiek2010 commented:
I tested this pr and noticed redundant logs after rotating my device:
![image](https://user-images.githubusercontent.com/3276264/60351593-634bd800-99c6-11e9-947a-4a36a3acdf72.png)

lognaturel commented:
Ahh, very good catch about the duplicate events, @grzesiek2010. I have `onStart` trigger an internal event called `activityDisplayed` but actually `onStart` gets called on rotation as well. That's annoying. I'll rename the event and I think I'll have to bring back the location permissions boolean. I thought I could get away without having it because I could tell if a user had navigated away from the app and came back but I didn't think about rotation going through the exact same lifecycle.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/logic/actions/setlocation/CollectSetLocationAction.java:
```diff
@@ -0,0 +1,100 @@
+        // Services are available. If it's not allowed, leave the target field blank.
+        if (GeneralSharedPreferences.getInstance().getBoolean(KEY_BACKGROUND_LOCATION, true)
+            && LocationClients.areGooglePlayServicesAvailable(Collect.getInstance().getApplicationContext())) {
+            maxAccuracyLocationClient.requestLocationUpdates(SECONDS_TO_CONSIDER_UPDATES);
```
I really like that the detail of these updates is now all hidden inside the client!

lognaturel replied:
💃 Thanks for talking through that with me!

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java:
```diff
@@ -0,0 +1,296 @@
+
+                return userMessage;
+
+            case RECEIVING_LOCATIONS:
```
Running `BackgroundLocationManagerTest` with coverage it doesn't look like this case gets hit by any of the tests

lognaturel replied:
Oh dear, that's because it was totally wrong and didn't represent a real case anymore. 😫 8b93f22 should address that.

seadowg commented on collect_app/src/test/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManagerTest.java:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    @Test
+    public void locationPermissions_ShouldBeRequested_WhenFormRequestsBackgroundLocation_AndPreconditionsAreMet() {
```
I found the test names here a little hard to read. Always hard to get to a readable state with JUnit methods but maybe they could be more focused with the method under test (like an `<subject>_<condition>_<expectation>` style) like:

`formFinishedLoading_whenPreconditionsAreMet_makesIsPendingPermissionCheckReturnTrue`

Maybe that's not better?

lognaturel replied:
🤷🏻‍♀️I think I was consistent with `<subject>_<expectation>_<most important conditions>`. I tried to make `subject` more about the thing that the user will experience.

seadowg commented on collect_app/src/test/java/org/odk/collect/android/location/client/MaxAccuracyWithinTimeoutLocationClientTest.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Test
+    public void requestingLocationUpdates_ShouldResultInUpdate() {
```
Tiny nitpick: Especially given the length of test names in JUnit I think we can leave out "should" (so just `requestingLocationUpdates_resultsInUpdate`)?

lognaturel replied:
Aw, I like my shoulds! But I see your point. c70eb98

seadowg commented:
I like the division of tests here. Basically it seems like you're using Espresso to drive out the users interactions and then driving out the detail of the background logging/tracking with unit tests around the responsible components.

I've found in the past I've often ended up with tests around my `ViewModel` but right now it's really thin so it makes sense that there isn't much to drive out there given what Espresso is covering.

lognaturel committed:
Describe and move the ViewModel

lognaturel committed:
Resume location capture after activity killed by system

In that case, the form doesn't need to be loaded so we don't get a form loaded event. If the form audits location, we should go straight to requesting locations or stopping if preconditions aren't met.

Removes the savedFormStart flag that was used to detect this case previously.

There is now one known limitation: if FormEntryActivity is killed while the hierarchy view is being shown after a saved form start, nothing related to background location will be displayed to the user or logged to the audit log when the activity does become visible.

lognaturel committed:
Log initialization events before question event

lognaturel committed:
Don't log Android location permissions change on orientation change

The activity is responsible for detecting changes in the permission because it needs to save the previous permission value across system-initiated death of the activity. In other words, we don't want to log a permission change just because the activity was killed due to low memory.

lognaturel committed:
Reuse no-accuracy variant in accuracy variant

lognaturel committed:
Make whether event type is logged or not a property of event types

lognaturel commented:
- [x] Consider using a `ViewModel` factory -- make sure dependencies aren't recreated on activity recreation https://github.com/opendatakit/collect/pull/3175#discussion_r298795931
- [ ] Consider introducing an interface for the helper/delegate https://github.com/opendatakit/collect/pull/3175#discussion_r298795696
- [x] Move `isInterval` to the `AuditEventType` enum https://github.com/opendatakit/collect/pull/3175#discussion_r298604022
- [x] Simplify logic for events that don't need to be logged https://github.com/opendatakit/collect/pull/3175#discussion_r299896816
- [x] Revisit test names https://github.com/opendatakit/collect/pull/3175#discussion_r299124080 and https://github.com/opendatakit/collect/pull/3175#discussion_r299125859

lognaturel committed:
Make whether event type is interval a property of event types

lognaturel committed:
Factor out logic to skip logging

lognaturel committed:
Build ViewModel dependencies in activity

In preparation for eventual ViewModel testing

lognaturel committed:
Shorten test names

lognaturel commented:
@seadowg and @grzesiek2010, thanks for the thorough review. I believe I have addressed all of your suggestions except for introducing an interface for the helper (see question at https://github.com/opendatakit/collect/pull/3175#discussion_r298795696).

I realized that I was not handling the "activity killed by system" case because `ViewModel`s are rebuilt then. To do that, I added 7ac5096. Hopefully the comments and commit message make that case clear.

There are two limitations that I'm aware of:
- If `do not save activities` is set, loading from a saved form will not show the location snackbar or write the location enabled message to the audit log. This is because the FormEntryActivity gets killed when the hierarchy view is displayed. I think this is a rare enough case and an acceptable enough "failure" mode that it's ok to leave as-is. If anyone disagrees strongly, I think we'd have to bring back the `savedFormStart` flag that was saved by `onSaveInstanceState`.
- If a form with a setlocation action triggered by form load is launched and triggers a permission request, the permission grant will come after the action has already been triggered so that blank form won't get a location. I don't think there's much that can be done about that so it should be documented.

@grzesiek2010, amazing that you were able to keep all these cases straight without tests and with everything tangled in the activity!!

I'm guessing there are still some edge cases I haven't considered so would be great to start QA soon!

seadowg commented:
@lognaturel all looks good. 

It's super confusing that `ViewModel` instances are preserved between "configuration changes" (which still calls `onDestroy()`) but not Activity destroys for reclaiming memory. I had a look into it and this blog explains how that all actually works under the hood: https://android.jlelse.eu/deep-dive-inside-of-androids-viewmodel-architecture-components-e6756dc0bb11.

I'd heard of using a "Holder Fragment"(using `setRetainInstance` ~evil~ magic) to preserve state between rotations etc but it's interesting to see Google are using that pattern themselves.

lognaturel commented:
> It's super confusing that ViewModel instances are preserved between "configuration changes" (which still calls onDestroy()) but not Activity destroys for reclaiming memory.

Good article, thanks. I'm really surprised that things were designed this way. Like you say, the approach feels hacky and worse is that it violates expectations. I'm going to continue advocating for any subclasses of the architecture components that we introduce being as thin as possible so that they can easily be replaced or rethought as new patterns emerge and/or Android moves on to the next shiny thing.

seadowg commented:
@lognaturel I'm into that. Generally it would be nice to have any significant logic in a service/helper/delegate/usecase etc rather than in a `ViewModel` and then it can just be the glue between your 
"Android" world and everything else.

lognaturel commented:
Since @seadowg has given a thumbs up and there's not a lot in the QA pipeline, I'm going to mark as `needs testing`. If there are cases I didn't consider, there might be additional work to do here.

lognaturel commented:
[setlocation-and-audit-location.xml.txt](https://github.com/opendatakit/collect/files/3374170/setlocation-and-audit-location.xml.txt) is a form with both the new setlocation feature and the existing location audit. I've added it to the PR description as well because the two features should work harmoniously together.

mmarciniak90 commented:
@lognaturel I want to ask about one behavior. I'm not sure that replacing the location is expected. I used form `setlocation-action-value-changed`. 

| Steps | Screenshots |
|---|---|
|1. I added the location to the second question manually and I saved submission. Added location is visible in hierarchy view| ![Screenshot_20190711-081746](https://user-images.githubusercontent.com/3907554/61028219-6f696980-a3b8-11e9-8fc4-ddcf497210bd.png) |
|2. I checked XML and both locations were visible | ![Screenshot_20190711-081806](https://user-images.githubusercontent.com/3907554/61028132-392bea00-a3b8-11e9-89c4-c0a077673c92.png)|
| 3. I opened this saved submission and I noticed that added **location was replaced on my current location**|  ![Screenshot_20190711-081815](https://user-images.githubusercontent.com/3907554/61028210-68daf200-a3b8-11e9-996f-c079384e9f2e.png) |

lognaturel commented:
Oh dear, I'm really sorry, @mmarciniak90, that was not the version of the form I wanted to share. I was trying to display the point captured in the background in a geopoint widget but I couldn't come up with a good way to do so. Ideally I'd like that display widget to be read-only but now if a geopoint is read-only the button isn't shown. So this form has a writeable geopoint widget with a value that always gets reset to your current location. That means the behavior you're observing is expected but I agree it's not a very useful form. 

In general, displaying the point that was collected in the background is a bit strange because the value can take a while to be set and then can be reset within the 20s timeout period. [setlocation-action-value-changed.xml.txt](https://github.com/opendatakit/collect/files/3384139/setlocation-action-value-changed.xml.txt) is what I meant to share -- it displays the text of the point. You should be able to swipe back and forth to it and see the accuracy get progressively better (lower).

You may also find it helpful to look at the logcat output filtered to the string "location" to see what's actually going on behind the scenes as you test the visible behavior.

mmarciniak90 commented:
I spend a lot of time testing this issue and I did not notice regressions but still, I cannot say that there are not any.

**Verified on Android:** 4.4, 5.1, 6.0, 7.0, 8.1

Compared with v1.23-beta

**Verified cases with odk:setlocation and Audit**
AUDIT events:
* location tracking enabled
* location tracking disabled
* location providers enabled
* location permissions granted
* location permissions not granted

AUDIT logs contain:
* END date for the first question
* Latitude
* Longitude
* Accuracy
* Tracked answers

**Verified cases with odk:setlocation**
* location is not saved if device-level permission is not given (Android 6.0, 7.0, 8.1)
* location tracking disabled 
* rotating
* activity killed by system (testing with 'Don't keep activities' turned on )
* in new form `setlocation-action-value-changed.xml`, location value is autofilled
* in form `setlocation-action-with-repeats.xml`- enabling and disabling the location tracking; location value is added to every repeat
* in form `setlocation-action-instance-load.xml`, user is moved to page with `Save Form and Exit` button but location is set in background. Value is visible in xml
* verified behaviors in form `setlocation-and-audit-location.xml`
* created and verified a form with tracking the changing of answers and odk:setlocation - 
[SetLocation_trackingChangesAndLocationTestForm.xml.txt](https://github.com/opendatakit/collect/files/3392257/SetLocation_trackingChangesAndLocationTestForm.xml.txt)


**General**
- location preference toggle is available from the overflow menu
- turning on/off the location preference toggle works as expected

I checked logs:
![Screenshot from 2019-07-15 13-21-20](https://user-images.githubusercontent.com/3907554/61212768-78817000-a703-11e9-9d39-20447c973e20.png)

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

lognaturel commented:
Thanks so much, @mmarciniak90! I know there's a lot here. But the good news is that since all of the background location behavior is now isolated, we won't risk regressions if we have to change things in `FormEntryActivity`.

@grzesiek2010 how about taking a last look to see if you think of any other cases and merging when you're ready so we can get a beta out?

grzesiek2010 commented:
> how about taking a last look to see if you think of any other cases and merging when you're ready so we can get a beta out?

I'll do that today

lognaturel committed:
Use JavaRosa hasAction method

It replaced hasSetlLocationAction after code review

lognaturel commented:
@grzesiek2010 I just force pushed to remove the JavaRosa JAR since master is now on the JR 15 snapshot. I did a clean rebase on master. The only commit that has actually changed is the very last one that used to pull in JR from JAR.

grzesiek2010 commented:
@lognaturel 
I noticed one issue: if I enable/disable GPS during filling a form I can't see logs in the audit file. On master, it adds
`location providers enabled`
`location providers disabled`

@mmarciniak90 did you notice the same?

mmarciniak90 commented:
@grzesiek2010 I agree. I was able to reproduce the mentioned problem.
I executed this case a few times and this information was added once. 
But most of the executions were devoid of this information. 
I'm not able to explain why it worked once.

lognaturel commented:
Thank you, @grzesiek2010! I have some ideas of what's going on. Worst case, the activity can log the provider change if needed. Let's merge this so we can start getting feedback on it and I'll file a follow-on issue.
